File: src/charting/jobs/AnimatedZoomJob.java
package com.github.mikephil.charting.jobs;

import android.animation.Animator;
import android.animation.ValueAnimator;
import android.annotation.SuppressLint;
import android.graphics.Matrix;
import android.view.View;

import com.github.mikephil.charting.charts.BarLineChartBase;
import com.github.mikephil.charting.components.YAxis;
import com.github.mikephil.charting.utils.ObjectPool;
import com.github.mikephil.charting.utils.Transformer;
import com.github.mikephil.charting.utils.ViewPortHandler;

/**
 * Created by Philipp Jahoda on 19/02/16.
 */
@SuppressLint("NewApi")
public class AnimatedZoomJob extends AnimatedViewPortJob implements Animator.AnimatorListener {

    private static ObjectPool<AnimatedZoomJob> pool;

    static {
        pool = ObjectPool.create(8, new AnimatedZoomJob(null,null,null,null,0,0,0,0,0,0,0,0,0,0));
    }

    public static AnimatedZoomJob getInstance(viewPortHandler: ViewPortHandler, View v, Transformer trans, YAxis axis, let xAxisRange, let scaleX, let scaleY, let xOrigin, let yOrigin, let zoomCenterX, let zoomCenterY, let zoomOriginX, let zoomOriginY, long duration) {
        AnimatedZoomJob result = pool.get();
        result.mViewPortHandler = viewPortHandler;
        result.xValue = scaleX;
        result.yValue = scaleY;
        result.mTrans = trans;
        result.view = v;
        result.xOrigin = xOrigin;
        result.yOrigin = yOrigin;
        result.yAxis = axis;
        result.xAxisRange = xAxisRange;
        result.resetAnimator();
        result.animator.setDuration(duration);
        return result;
    }

    protected let zoomOriginX;
    protected let zoomOriginY;

    protected let zoomCenterX;
    protected let zoomCenterY;

    protected YAxis yAxis;

    protected let xAxisRange;

    @SuppressLint("NewApi")
    public AnimatedZoomJob(viewPortHandler: ViewPortHandler, View v, Transformer trans, YAxis axis, let xAxisRange, let scaleX, let scaleY, let xOrigin, let yOrigin, let zoomCenterX, let zoomCenterY, let zoomOriginX, let zoomOriginY, long duration) {
        super(viewPortHandler, scaleX, scaleY, trans, v, xOrigin, yOrigin, duration);

        this.zoomCenterX = zoomCenterX;
        this.zoomCenterY = zoomCenterY;
        this.zoomOriginX = zoomOriginX;
        this.zoomOriginY = zoomOriginY;
        this.animator.addListener(this);
        this.yAxis = axis;
        this.xAxisRange = xAxisRange;
    }

    protected Matrix this.mOnAnimationUpdateMatrixBuffer = new Matrix();
    
    public onAnimationUpdate(ValueAnimator animation) {

        let scaleX = xOrigin + (xValue - xOrigin) * phase;
        let scaleY = yOrigin + (yValue - yOrigin) * phase;

        Matrix save = this.mOnAnimationUpdateMatrixBuffer;
        this.mViewPortHandler.setZoom(scaleX, scaleY, save);
        this.mViewPortHandler.refresh(save, view, false);

        let valsInView = yAxis.mAxisRange / this.mViewPortHandler.getScaleY();
        let xsInView =  xAxisRange / this.mViewPortHandler.getScaleX();

        pts[0] = zoomOriginX + ((zoomCenterX - xsInView / 2f) - zoomOriginX) * phase;
        pts[1] = zoomOriginY + ((zoomCenterY + valsInView / 2f) - zoomOriginY) * phase;

        this.mTrans.pointValuesToPixel(pts);

        this.mViewPortHandler.translate(pts, save);
        this.mViewPortHandler.refresh(save, view, true);
    }

    
    public onAnimationEnd(Animator animation) {
        ((BarLineChartBase) view).calculateOffsets();
        view.postInvalidate();
    }

    
    public onAnimationCancel(Animator animation) {

    }

    
    public onAnimationRepeat(Animator animation) {

    }

    
    public recycleSelf() {

    }

    
    public onAnimationStart(Animator animation) {

    }

    
    protected ObjectPool.Poolable instantiate() {
        return new AnimatedZoomJob(null,null,null,null,0,0,0,0,0,0,0,0,0,0);
    }
}

File: src/charting/jobs/AnimatedZoomJob.ts
import { Matrix } from '@nativescript-community/ui-canvas';
import { BarLineChartBase } from '../charts/BarLineChartBase';
import { YAxis } from '../components/YAxis';
import { ObjectPool } from '../utils/ObjectPool';
import { Transformer } from '../utils/Transformer';
import { Utils } from '../utils/Utils';
import { ViewPortHandler } from '../utils/ViewPortHandler';
import { AnimatedViewPortJob } from './AnimatedViewPortJob';
/**
 * Created by Philipp Jahoda on 19/02/16.
 */
export class AnimatedZoomJob extends AnimatedViewPortJob {
    protected mZoomOriginX: number;
    protected mZoomOriginY: number;

    protected mZoomCenterX: number;
    protected mZoomCenterY: number;

    protected mYAxis: YAxis;

    protected mXAxisRange;

    public static getInstance(
        viewPortHandler: ViewPortHandler,
        v: BarLineChartBase<any, any, any>,
        trans: Transformer,
        axis: YAxis,
        xAxisRange,
        scaleX: number,
        scaleY: number,
        xOrigin: number,
        yOrigin: number,
        zoomCenterX: number,
        zoomCenterY: number,
        zoomOriginX: number,
        zoomOriginY: number,
        duration: number
    ) {
        const result = pool.get();
        result.mViewPortHandler = viewPortHandler;
        result.mXValue = scaleX;
        result.mYValue = scaleY;
        result.mTrans = trans;
        result.mView = v;
        result.mXOrigin = xOrigin;
        result.mYOrigin = yOrigin;

        result.mZoomCenterX = zoomCenterX;
        result.mZoomCenterY = zoomCenterY;
        result.mZoomOriginX = zoomOriginX;
        result.mZoomOriginY = zoomOriginY;

        result.mYAxis = axis;
        result.mXAxisRange = xAxisRange;
        result.createAnimator(duration);
        return result;
    }

    constructor(
        viewPortHandler: ViewPortHandler,
        v: BarLineChartBase<any, any, any>,
        trans: Transformer,
        axis: YAxis,
        xAxisRange,
        scaleX: number,
        scaleY: number,
        xOrigin: number,
        yOrigin: number,
        zoomCenterX: number,
        zoomCenterY: number,
        zoomOriginX: number,
        zoomOriginY: number,
        duration: number
    ) {
        super(viewPortHandler, scaleX, scaleY, trans, v, xOrigin, yOrigin, duration);

        this.mZoomCenterX = zoomCenterX;
        this.mZoomCenterY = zoomCenterY;
        this.mZoomOriginX = zoomOriginX;
        this.mZoomOriginY = zoomOriginY;
        // this.animator.addListener(this);
        this.mYAxis = axis;
        this.mXAxisRange = xAxisRange;
    }

    public onAnimationUpdate(animation) {
        const scaleX = this.mXOrigin + (this.mXValue - this.mXOrigin) * this.mPhase;
        const scaleY = this.mYOrigin + (this.mYValue - this.mYOrigin) * this.mPhase;

        const save = Utils.getTempMatrix();
        const viewPortHanlder = this.mViewPortHandler;
        viewPortHanlder.setZoom(scaleX, scaleY, save);
        viewPortHanlder.refresh(save, this.mView, false);

        const valsInView = this.mYAxis.mAxisRange / viewPortHanlder.getScaleY();
        const xsInView = this.mXAxisRange / viewPortHanlder.getScaleX();
        const pts = Utils.getTempArray(2);

        pts[0] = this.mZoomOriginX + (this.mZoomCenterX - xsInView / 2 - this.mZoomOriginX) * this.mPhase;
        pts[1] = this.mZoomOriginY + (this.mZoomCenterY + valsInView / 2 - this.mZoomOriginY) * this.mPhase;

        this.mTrans.pointValuesToPixel(pts);

        viewPortHanlder.translate(pts, save);
        viewPortHanlder.refresh(save, this.mView, true);
    }

    public onAnimationEnd(animation) {
        this.mView.calculateOffsets();
        this.mView.invalidate();
    }

    public onAnimationCancel(animation) {}

    public recycleSelf() {}

    public instantiate() {
        return new AnimatedZoomJob(null, null, null, null, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0);
    }
}
const pool = ObjectPool.create<AnimatedZoomJob>(8, new AnimatedZoomJob(null, null, null, null, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)).setReplenishPercentage(0.5);
